packages/plugin-ext-vscode/src/browser/plugin-vscode-commands-contribution.ts

```diff
@@ -132,29 +136,82 @@ export class PluginVscodeCommandsContribution implements CommandContribution {
+
+        let opener: OpenHandler | undefined;
+        if (typeof viewTypeId === 'string') {
+            const lowerViewType = viewTypeId.toLowerCase();
+            const openers = await this.openerService.getOpeners();
+            for (const opnr of openers) {
+                const idLowerCase = opnr.id.toLowerCase();
+                if (lowerViewType === idLowerCase) {
+                    opener = opnr;
+                    break;
+                }
+            }
+        } else {
+            opener = await this.openerService.getOpener(uri, editorOptions);
+        }
+
+        if (opener) {
+            await opener.open(uri, editorOptions);
+            return true;
+        }
+
+        return false;
+    }
+
     registerCommands(commands: CommandRegistry): void {
         commands.registerCommand(VscodeCommands.OPEN, {
             isVisible: () => false,
             execute: async (resource: URI, columnOrOptions?: ViewColumn | TextDocumentShowOptions) => {
-                if (!resource) {
-                    throw new Error(`${VscodeCommands.OPEN.id} command requires at least URI argument.`);
+                const result = await this.openWith(VscodeCommands.OPEN.id, resource, columnOrOptions);
+                if (!result) {
+                    throw new Error(`Could not find an editor for ${resource}`);
                 }
-                if (!URI.isUri(resource)) {
-                    throw new Error(`Invalid argument for ${VscodeCommands.OPEN.id} command with URI argument. Found ${resource}`);
+            }
+        });
+
+        commands.registerCommand(VscodeCommands.OPEN_WITH, {
+            isVisible: () => false,
+            execute: async (resource: URI, viewTypeId: string, columnOrOptions?: ViewColumn | TextDocumentShowOptions) => {
+                if (!viewTypeId) {
+                    throw new Error(`Running the contributed command: ${VscodeCommands.OPEN_WITH} failed.`);
                 }
 
-                let options: TextDocumentShowOptions | undefined;
-                if (typeof columnOrOptions === 'number') {
-                    options = {
-                        viewColumn: columnOrOptions
-                    };
-                } else if (columnOrOptions) {
-                    options = {
-                        ...columnOrOptions
-                    };
+                if (viewTypeId.toLowerCase() === 'default') {
+                    return commands.executeCommand(VscodeCommands.OPEN.id, resource, columnOrOptions);
+                }
+
+                let result = await this.openWith(VscodeCommands.OPEN_WITH.id, resource, columnOrOptions, viewTypeId);
+                if (!result) {
+                    // Theia set custom editor id using 'custom-editor-' prefix
+                    result = await this.openWith(VscodeCommands.OPEN_WITH.id, resource, columnOrOptions, `custom-editor-${viewTypeId}`);
```

Comment:
I think this only resolves half of the issues @tsmaeder had with this approach. Using `custom-editor-${viewTypeId}` leaks internals of the custom editor to other code. Please add something like the following code to the end of `custom-editor-opener.tsx` and use the method here and in the constructor of the `CustomEditorOpener` as well:

```ts
export namespace CustomEditorOpener {
    export function toCustomEditorId(editorViewType: string): string {
        return `custom-editor-${editorViewType}`;
    }
}
```

Reply:
That's an appropriate point.
This part is implemented differently than VSCode and Theia.
We will consider this carefully.

Reply:
@msujew After reviewing what you pointed out, I found that CustomEditor always has a 'custom-editor-' prefix by default.
So I modified the code to always prefix the viewType with that.

Reply:
@msujew I missed something in your comments.
I modified the code again to add it.
Please check it